zebrad/tests/acceptance.rs

```diff
@@ -324,57 +351,84 @@ fn persistent_mode() -> Result<()> {
     // Make sure the command was killed
     output.assert_was_killed()?;
 
-    // Check that we have persistent rocksdb database
     let cache_dir = testdir.path().join("state");
-    assert_with_context!(cache_dir.read_dir()?.count() > 0, &output);
+    assert_with_context!(
+        cache_dir.read_dir()?.count() > 0,
+        &output,
+        "unexpected empty state directory from persistent state config"
+    );
 
     Ok(())
 }
 
-#[test]
-fn ephemeral_mode() -> Result<()> {
-    zebra_test::init();
-
-    let testdir = testdir()?.with_config(default_test_config()?)?;
-    let testdir = &testdir;
+/// The cache_dir config used in the ephemeral mode tests
+#[derive(Debug, PartialEq, Eq)]
+enum EphemeralConfig {
+    /// the cache_dir config is left at its default value
+    Default,
+    /// the cache_dir config is set to a path in the tempdir
+    MisconfiguredCacheDir,
+}
 
-    // Any free argument is valid
-    let mut child = testdir.spawn_child(&["start", "argument"])?;
-    // Run the program and kill it after a few seconds
-    std::thread::sleep(LAUNCH_DELAY);
-    child.kill()?;
-    let output = child.wait_with_output()?;
+/// The check performed by the ephemeral mode tests
+#[derive(Debug, PartialEq, Eq)]
+enum EphemeralCheck {
+    /// an existing directory is not deleted
+    ExistingDirectory,
+    /// a missing directory is not created
+    MissingDirectory,
+}
 
-    // Make sure the command was killed
-    output.assert_was_killed()?;
+#[test]
+fn ephemeral_existing_directory() -> Result<()> {
+    ephemeral(EphemeralConfig::Default, EphemeralCheck::ExistingDirectory)
+}
 
-    let cache_dir = testdir.path().join("state");
-    assert_with_context!(!cache_dir.exists(), &output);
+#[test]
+fn ephemeral_missing_directory() -> Result<()> {
+    ephemeral(EphemeralConfig::Default, EphemeralCheck::MissingDirectory)
+}
 
-    Ok(())
+#[test]
+fn misconfigured_ephemeral_existing_directory() -> Result<()> {
+    ephemeral(
+        EphemeralConfig::MisconfiguredCacheDir,
+        EphemeralCheck::ExistingDirectory,
+    )
 }
 
 #[test]
-fn misconfigured_ephemeral_mode() -> Result<()> {
-    zebra_test::init();
+fn misconfigured_ephemeral_missing_directory() -> Result<()> {
+    ephemeral(
+        EphemeralConfig::MisconfiguredCacheDir,
+        EphemeralCheck::MissingDirectory,
+    )
+}
 
-    let dir = TempDir::new("zebrad_tests")?;
-    let cache_dir = dir.path().join("state");
-    fs::create_dir(&cache_dir)?;
+fn ephemeral(cache_dir_config: EphemeralConfig, cache_dir_check: EphemeralCheck) -> Result<()> {
```

Comment:
It would be really cool if we could somehow make a system for generating test matrixes in situations like this. Like, given a set of enums that define each dimension and a test fn that evaluates a point in that enum combination space automatically create and run a version of that test for every combination of inputs.

Reply:
Yeah, that would be a really neat feature.

We'll need `IterVariants` from the `enum_derive` crate to iterate through the full list of enum values:
https://docs.rs/enum_derive/0.1.7/enum_derive/index.html#overview

Reply:
Did you want to open a follow-up ticket for this?
Or look around for a crate that already implements it?

I think you have the best idea of what you want here :-)

Reply:
Not really sure, took a look at `enum_derive` but wasn't super impressed with it. I don't know of another tool for this so I feel the solution might be to just make our own proc macro crate for it. I'll create a follow up issue for this. I think it might be another good issue for outside contributors.

Reply:
@yaahc did you create an issue for the test matrix follow-up?